### PR TITLE
[READY]SDQL2 update: In which admins get more power than they deserve.

### PR DESCRIPTION
**Pull request by silicons**
#23694 Credits to /vg/station and @PJB3005
:cl:
rscadd: ADMINS: SDQL2 has been given some new features!
bugfix: SDQL2 now gives you an exception on runtime instead of flooding server runtime logs.
rscadd: SDQL2 now supports usr, which makes that variable reference to whatever mob you are in, src, which targets the object it is being called on itself, and marked, which targets the datum marked by the admin calling it. Also, it supports hex references (the hex number at the top of a VV panel) in {}s, so you can target nearly anything! Also, global procs are supported by global.[procname](args), for CALL queries.
bugfix: SDQL2 can no longer edit /datum/admins or /datum/admin_rank, and is protected from changing x/y/z of a turf and anything that would cause broken movement for movable atoms.
rscadd: SDQL2 can now get list input with [arg1, arg2]!
experimental: Do '"<string>"' to put strings inside of SDQL2 or it won't work.
/:cl:

### CODE REVIEW

**Comment by PJB3005** [comment]
Real list of changes:

SDQL can do `SET` on nested objects.
SDQL call proc has been fixed. Correctly handles arguments now.
You can call any proc inside SDQL expressions.
`src`, `usr`, `marked` and `global` variables added:
* `src` is the current object selected by the query.
* `usr` is you, the admin.
* `marked` is your admin marked datum
* `global` is a token for global variables and procs. (seems the former bit isn't ported?)
* `std_` procs to workaround lack of calling builtins.
* List access with `list[expression]`
* List expressions: `[expr, expr, ...]` or and assoc list support: `[expr: expr, expr: expr, ...]`
* `{0xdeadbeef}` for references.
* `IN` now allows using arbitrary expressions. For example: `SELECT /mob IN global.living_mob_list`
* Probably some other shit I forgot.

File: code/game/turfs/turf.dm [hunk 1]
```diff
@@ -21,9 +21,14 @@
 
 	var/explosion_level = 0	//for preventing explosion dodging
 	var/explosion_id = 0
-	
+
 	var/list/decals
 
+/turf/SDQL_update(var/list/vars_list)
```
Comment (RemieRichards) [inline code/game/turfs/turf.dm]: don't do the ```var/``` in proc arguments, in this case ```list/var_list``` is all that's needed.

Comment (kevinz000) [inline code/game/turfs/turf.dm]: K

**Comment by kevinz000** [comment]
@PJB3005 some of them aren't put in yet, like nested objects and whatnot D:
only did some of them.

**Comment by kevinz000** [comment]
as in I didn't want to generate a global var list when I've already been told that some of them are not accessible for a good reason.

**Comment by RemieRichards** [comment]
@kevinz000 you know damn well I meant the entire PR and not that one line.

**Comment by kevinz000** [comment]
@RemieRichards Oh.

**Comment by kevinz000** [comment]
@RemieRichards Is catch a proc..?

**Comment by PJB3005** [comment]
Why don't you just remove the current files you have and then replace them directly with the /vg/ code?

**Comment by RemieRichards** [comment]
@kevinz000 catch is a keyword but it should be done there too, yes.

**Comment by kevinz000** [comment]
@PJB3005 Wouldn't that end badly since the codebases are different and copypasting from one to another without even looking at the code is normally a bad thing?

**Comment by ExcessiveUseOfCobblestone** [comment]
WHO DELETED MY POST

**Comment by kevinz000** [comment]
@ExcessiveUseOfCobblestone made new pr since i decided making two prs was dumb.

**Comment by kevinz000** [comment]
while i'm here I might as well port the resto f the stuff.
and not just copypaste it :P

**Comment by PKPenguin321** [comment]
extreme changelog bloat. condense it to at the absolute most 3 lines. it's questionable if it's even needed in the first place considering it's admin-only, as well.

**Comment by kevinz000** [comment]
you're implying admins will ever notice it otherwise , and that any of us knows how to read the sdql code.

**Comment by PKPenguin321** [comment]
your changelog does not need to be 50 lines.

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm [hunk 2]
```diff
@@ -389,6 +396,12 @@
 	else if(copytext(expression[i], 1, 2) in list("'", "\""))
 		val = copytext(expression[i], 2, length(expression[i]))
 
+	else if(expression[i] == "\[")
+		var/list/expressions_list = expression[++i]
+		val = list()
+		for(var/list/expression_list in expressions_list)
+			val += SDQL_expression(object, expression_list)
```
Comment (PJB3005) [inline code/modules/admin/verbs/SDQL2/SDQL_2.dm]: It just occurred to me that nested lists probably don't work.

Make this `val[++val.len] =` instead.

**Comment by kevinz000** [comment]
@PJB3005 Doesn't work.

**Comment by PJB3005** [comment]
Oh eight the variable needs to be cast into a list first.

File: code/modules/admin/verbs/SDQL2/SDQL_2.dm [hunk 3]
```diff
@@ -18,11 +18,22 @@
 
 */
 
+/datum/proc/SDQL_update(const/var_name, new_value)
+	vars[var_name] = new_value
+	return TRUE
```
Comment (MrStonedOne) [inline code/modules/admin/verbs/SDQL2/SDQL_2.dm]: This really should have been using the existing vv_edit_var() proc.

**Pull request merged**